File: Leetcode/Python/2176. Count Equal and Divisible Pairs in an Array/solution.py
# https://leetcode.com/problems/count-equal-and-divisible-pairs-in-an-array/description

class Solution:
    def countPairs(self, nums: List[int], k: int) -> int:
        pairs = 0
        hash = dict()

        for idx, num in enumerate(nums):
            if num not in hash:
                hash[num] = []
            else:
                for i in hash[num]:
                    if (i * idx) % k == 0:
                        pairs += 1
            
            hash[num].append(idx)

        return pairs

File: Leetcode/Typescript/2176. Count Equal and Divisible Pairs in an Array/solution.ts
// https://leetcode.com/problems/count-equal-and-divisible-pairs-in-an-array/description

function countPairs(nums: number[], k: number): number {
  const indexHash = new Map<number, number[]>();
  let pairs = 0;

  for (let idx = 0; idx < nums.length; idx++) {
    const key = nums[idx];
    if (!indexHash.has(key)) {
      indexHash.set(key, []);
    } else {
      for (const i of indexHash.get(key)!) {
        if ((idx * i) % k === 0) {
          pairs += 1;
        }
      }
    }

    indexHash.get(key)!.push(idx);
  }

  return pairs;
}
